subprojects/execution/src/test/groovy/org/gradle/internal/execution/impl/steps/IncrementalExecutionTest.groovy

```diff
@@ -0,0 +1,821 @@
+import org.gradle.internal.snapshot.impl.ImplementationSnapshot
+import org.gradle.test.fixtures.file.TestFile
+import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
+import org.gradle.testing.internal.util.Specification
+import org.junit.Rule
+
+import java.time.Duration
+import java.util.function.BooleanSupplier
+
+class IncrementalExecutionTest extends Specification {
+
+    @Rule
+    final TestNameTestDirectoryProvider temporaryFolder = TestNameTestDirectoryProvider.newInstance()
+
+    def fileHasher = new TestFileHasher()
+    def fileSystemMirror = new DefaultFileSystemMirror(Stub(WellKnownFileLocations))
+    def snapshotter = new DefaultFileSystemSnapshotter(fileHasher, new StringInterner(), TestFiles.fileSystem(), fileSystemMirror)
+    def fingerprinter = new AbsolutePathFileCollectionFingerprinter(new StringInterner(), snapshotter)
+    def outputFingerprinter = new OutputFileCollectionFingerprinter(new StringInterner(), snapshotter)
+    def executionHistoryStore = new TestExecutionHistoryStore()
+    def outputChangeListener = new OutputChangeListener() {
+        @Override
+        void beforeOutputChange() {
+            fileSystemMirror.beforeOutputChange()
+        }
+
+        @Override
+        void beforeOutputChange(Iterable<String> affectedOutputPaths) {
+            fileSystemMirror.beforeOutputChange(affectedOutputPaths)
+        }
+    }
+    def outputFilesRepository = new TestOutputFilesRepository()
+    def buildInvocationScopeId = new BuildInvocationScopeId(UniqueId.generate())
+    def classloaderHierarchyHasher = new ClassLoaderHierarchyHasher() {
+        @Override
+        HashCode getClassLoaderHash(ClassLoader classLoader) {
+            return HashCode.fromInt(1234)
+        }
+    }
+    def valueSnapshotter = new DefaultValueSnapshotter(classloaderHierarchyHasher, new NamedObjectInstantiator())
+
+    final outputFile = temporaryFolder.file("output-file")
+    final outputDir = temporaryFolder.file("output-dir")
+    final outputDirFile = outputDir.file("some-file")
+    final outputDirFile2 = outputDir.file("some-file-2")
+    final emptyOutputDir = temporaryFolder.file("empty-output-dir")
+    final missingOutputFile = temporaryFolder.file("missing-output-file")
+    final missingOutputDir = temporaryFolder.file("missing-output-dir")
+    final inputFile = temporaryFolder.createFile("input-file")
+    final inputDir = temporaryFolder.createDir("input-dir")
+    final inputDirFile = inputDir.file("input-file2").createFile()
+    final missingInputFile = temporaryFolder.file("missing-input-file")
+    final inputFiles = [file: [inputFile], dir: [inputDir], missingFile: [missingInputFile]]
+    final outputFiles = [file: [outputFile], missingFile: [missingOutputFile]]
+    final outputDirs = [emptyDir: [emptyOutputDir], dir: [outputDir], missingDir: [missingOutputDir]]
+    final createFiles = [outputFile, outputDirFile, outputDirFile2] as Set
```

Comment:
We could extract the whole output/input file property structure into another test fixture. I would leave it as is for now.

Do you feel strongly about it?

Reply:
A fixture would be a different direction. The point of that would be to have a better documented _thing_ that helps you test incremental execution. What I'm suggesting would rely on the fact that most of the tests need only a small subset of these declarations, so it would be obvious what is being relied on in each test. That would not be the case with a fixture.

Reply:
I'm fine leaving it like this for now, too.